tools/generator/dfg/sam/sam_device_tree.py

```diff
@@ -0,0 +1,244 @@
+            interrupts.append({'position': i.get('index'), 'name': i.get('name')})
+        p['interrupts'] = interrupts
+
+        #### Events are similar to interrupts, but instead of triggering an ISR data is passed from source
+        #### to sink without waking the processor
+        # event sources
+        event_sources = []
+        for i in device_file.query("//events/generators/generator"):
+            event_sources.append({'index': i.get('index'), 'name': i.get('name'), 'instance': i.get('module-instance')})
+        p['event_sources'] = event_sources
+
+        # event sinks or 'users' as the datasheet calls them
+        event_users = []
+        for i in device_file.query("//events/users/user"):
+            event_users.append({'index': i.get('index'), 'name': i.get('name'), 'instance': i.get('module-instance')})
+        p['event_users'] = event_users
+
+        LOGGER.debug("Found GPIOs: [%s]", ", ".join([p.upper() + i for p,i in p['gpios']]))
+        LOGGER.debug("Available Modules are:\n" + SAMDeviceTree._modulesToString(p['modules']))
+        # LOGGER.debug("Found Signals:")
+        # for sig in p['signals']:
+        #     LOGGER.debug("    %s", sig)
+        # LOGGER.debug("Found Interrupts:")
+        # for intr in p['interrupts']:
+        #     LOGGER.debug("    %s", intr)
+        # LOGGER.debug("Found Event Sources:")
+        # for eventSrc in p['event_sources']:
+        #     LOGGER.debug("    %s", eventSrc)
+        # LOGGER.debug("Found Event Users:")
+        # for eventUsr in p['event_users']:
+        #     LOGGER.debug("    %s", eventUsr)
+
+        return p
+
+    @staticmethod
+    def _modulesToString(modules):
+        string = ""
+        mods = sorted(modules)
+        char = mods[0][0][0:1]
+        for module, instance in mods:
+            if not instance.startswith(char):
+                string += "\n"
+            string += instance + " \t"
+            char = instance[0][0:1]
+        return string
+
+    @staticmethod
+    def _device_tree_from_properties(p):
+        tree = DeviceTree('device')
+        tree.ids.append(p['id'])
+
+        def topLevelOrder(e):
+            order = ['attribute-flash', 'attribute-ram', 'attribute-eeprom', 'attribute-core', 'attribute-mcu', 'header', 'attribute-define']
+            if e.name in order:
+                if e.name in ['attribute-flash', 'attribute-eeprom', 'attribute-ram']:
+                    return (order.index(e.name), int(e['value']))
+                else:
+                    return (order.index(e.name), e['value'])
+            return (len(order), -1)
+        tree.addSortKey(topLevelOrder)
+
+        # SAMDeviceTree.addDeviceAttributesToNode(p, tree, 'attribute-flash')
+        # SAMDeviceTree.addDeviceAttributesToNode(p, tree, 'attribute-ram')
+        # SAMDeviceTree.addDeviceAttributesToNode(p, tree, 'attribute-eeprom')
+        SAMDeviceTree.addDeviceAttributesToNode(p, tree, 'attribute-mcu')
```

Comment:
The `topLevelOrder` can be removed, same with `attribute-mcu`. They are left over from the AVRs.